library/network/src/modules/SuSEFirewall.rb

```diff
@@ -154,9 +154,64 @@ def self.create
   # ----------------------------------------------------------------------------
   # SuSEFirewalld Class. Trying to provide relevent pieces of SF2 functionality via
   # firewalld.
-  class SuSEFirewalld < Module
+  class SuSEFirewalld < Firewall
+    attr_reader :special_all_interface_zone
+
+    # Valid attributes for firewalld zones
+    @@zone_attributes=[:interfaces, :masquerade, :modified, :ports, :protocols, :services]
+    # {enable,start}_firewall are "inherited" from SF2 so we can't use symbols
+    # there without having to change all the SF2 callers.
+    @@key_settings=["enable_firewall", :logging, "start_firewall"]
+
+    # We need that for the tests. Nothing else should access the API
+    # directly
+    def api
+      @fwd_api
+    end
+
     def initialize
+      # firewalld API interface.
+      @fwd_api = FirewalldAPI.create
+      # firewalld service
+      @firewall_service = "firewalld"
+      # firewalld package
+      @FIREWALL_PACKAGE = "firewalld"
+      # firewall settings map
+      @SETTINGS = {}
+      # Zone which works with the special_all_interface_string string. In our case,
+      # we don't want to deal with this just yet. FIXME
+      @special_all_interface_zone = ""
+    end
+
+    # Function which attempts to convert a sf2_service name to a firewalld
+    # equivalent.
+    def _sf2_to_firewalld_service(service)
+
+      # First, let's strip off 'service:' from service name if present.
+      tmp_service = service.include?("service:") ?
+        service.partition(":")[2] : service
+
+      sf2_to_firewalld_map = {
+        # netbios is covered in the samba service file
+        "netbios-server" => ["samba"],
+        "nfs-client" => ["nfs"],
+        "nfs-kernel-server" => ["mountd", "nfs", "rpc-bind"],
+        "samba-server" => ["samba"],
+        "sshd" => ["ssh"]
+      }
+
+      if sf2_to_firewalld_map.has_key?(tmp_service)
+        deep_copy(sf2_to_firewalld_map[tmp_service])
+      else
+        deep_copy([tmp_service])
```

Comment:
Neither of these occurrences of `deep_copy` is needed.

It is a bit tricky: In Ruby, variables are bindings to mutable objects, but in YCP all data was immutable. So when we converted the code, we added all these [deep_copy](https://github.com/yast/yast-ruby-bindings/blob/b73d72e8e1634e24b2671fd33019d4798ef8210a/src/ruby/yast/yast.rb#L60) calls to make copies of the arrays and hashes. Initially we thought we should also be copying strings. But it made the code so awfully slow. And we realized that all the YCP functions that we use make a copy of a string argument anyway.

As a result: **`deep_copy` makes deep copies of Hashes and Arrays, but does NOT copy the Strings.**

in the body of this method, we make new arrays that we are going to return, so `deep_copy` is not needed for that.
OTOH we may return the *service* argument if we did not find a translation, so you may be tempted to use `deep_copy` to get a new string. But it does not work that way.